File: ainara/framework/capabilities/mcp.py
# Ainara AI Companion Framework Project
# Copyright (C) 2025 Rubén Gómez - khromalabs.org
#
# This file is dual-licensed under:
# 1. GNU Lesser General Public License v3.0 (LGPL-3.0)
#    (See the included LICENSE_LGPL3.txt file or look into
#    <https://www.gnu.org/licenses/lgpl-3.0.html> for details)
# 2. Commercial license
#    (Contact: rgomez@khromalabs.org for licensing options)
#
# You may use, distribute and modify this code under the terms of either license.
# This notice must be preserved in all copies or substantial portions of the code.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.

import logging
from typing import Any, Dict

from ainara.framework.mcp.client_manager import MCPClientManager
from ainara.framework.mcp.tool import MCPTool

from .base import CapabilityProvider

logger = logging.getLogger(__name__)


class MCPToolProvider(CapabilityProvider):
    """Provider for discovering and executing MCP tools."""

    def __init__(self, mcp_client_manager: MCPClientManager):
        self.mcp_client_manager = mcp_client_manager
        self.capabilities: Dict[str, Dict[str, Any]] = {}

    def discover(self) -> Dict[str, Dict[str, Any]]:
        """Discover MCP tools and add them to the capabilities dictionary."""
        self.capabilities = {}
        if not self.mcp_client_manager:
            logger.info("MCP Client Manager not available, skipping MCP tool discovery.")
            return {}

        logger.info("Discovering MCP tools...")
        discovered_count = 0
        try:
            mcp_tools = self.mcp_client_manager.connect_and_discover()
            for tool in mcp_tools:
                capability_info = {
                    "instance": tool,
                    "type": "mcp",
                    "origin": "remote",
                    "description": tool.description,
                    "server": tool.server_name,
                    "hidden": False,
                    "run_info": self._get_mcp_tool_details(tool),
                }
                self.capabilities[tool.prefixed_name] = capability_info
                discovered_count += 1
                logger.info(
                    f"Discovered MCP tool: {tool.prefixed_name} from"
                    f" {tool.server_name}"
                )
        except Exception as e:
            logger.error(f"Failed during MCP tool discovery: {e}", exc_info=True)

        logger.info(f"Discovered {discovered_count} MCP tools.")
        return self.capabilities

    def execute(self, name: str, arguments: Dict[str, Any]) -> Any:
        """Execute an MCP tool."""
        if not self.mcp_client_manager:
            raise RuntimeError("MCP Client Manager not available for execution.")

        logger.info(f"Executing MCP tool: {name} with args: {arguments}")
        try:
            return self.mcp_client_manager.execute_tool(name, arguments)
        except Exception as e:
            logger.error(f"Error executing MCP tool '{name}': {e}", exc_info=True)
            raise RuntimeError(f"Failed to execute MCP tool '{name}': {e}") from e

    def format_for_llm(self, capability_data: Dict[str, Any]) -> str:
        """Format an MCP tool's description for an LLM prompt."""
        instance = capability_data["instance"]
        if hasattr(instance, "format_for_llm"):
            return instance.format_for_llm() + "---\n"

        # Fallback formatting
        name = next(
            (k for k, v in self.capabilities.items() if v == capability_data), None
        )
        if not name:
            return ""

        run_info = capability_data.get("run_info", {})
        params = run_info.get("parameters", {})
        desc = f"Tool: {name}\n"
        desc += f"Description: {capability_data['description']}\n"
        desc += f"Server: {capability_data['server']}\n"
        if params:
            desc += "Arguments:\n"
            for p_name, p_info in params.items():
                desc += f"- {p_name} (type: {p_info['type']})"
                if p_info["required"]:
                    desc += " (required)"
                desc += f": {p_info['description']}\n"
        return desc + "---\n"

    def _get_mcp_tool_details(self, tool: MCPTool) -> Dict[str, Any]:
        """Format MCP tool details similar to native skill run_info."""
        details = {
            "description": (
                f"Executes the MCP tool '{tool.name}' on server"
                f" '{tool.server_name}'."
            ),
            "parameters": {},
            "return_type": "any",
            "error": None,
        }
        try:
            if "properties" in tool.input_schema:
                required_params = tool.input_schema.get("required", [])
                for param_name, param_schema in tool.input_schema[
                    "properties"
                ].items():
                    details["parameters"][param_name] = {
                        "type": param_schema.get("type", "any"),
                        "default": "None",
                        "required": param_name in required_params,
                        "description": param_schema.get(
                            "description", "No description provided."
                        ),
                    }
        except Exception as e:
            logger.error(
                "Error parsing input schema for MCP tool"
                f" '{tool.prefixed_name}': {e}",
                exc_info=True,
            )
            details["error"] = f"Failed to parse input schema: {e}"
        return details

File: polaris/components/setup/steps/mcp.js
// Ainara AI Companion Framework Project
// Copyright (C) 2025 Rubén Gómez - khromalabs.org
//
// This file is dual-licensed under:
// 1. GNU Lesser General Public License v3.0 (LGPL-3.0)
//    (See the included LICENSE_LGPL3.txt file or look into
//    <https://www.gnu.org/licenses/lgpl-3.0.html> for details)
// 2. Commercial license
//    (Contact: rgomez@khromalabs.org for licensing options)
//
// You may use, distribute and modify this code under the terms of either license.
// This notice must be preserved in all copies or substantial portions of the code.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// Lesser General Public License for more details.

module.exports = {
    id: 'mcp',

    async init(ctx) {
        await generateMcpUI(ctx);
    },

    async save(ctx) {
        await saveMcpConfig(ctx);
    },

    validate(ctx) {
        return validateMcpStep(ctx);
    },

    addServerForm(ctx, serverName, container) {
        addMcpServerForm(ctx, serverName, container);
    }
};

// Copy the following functions from setup.js into this module,
// adapting them to use ctx.api, ctx.config, ctx.modifiedFields, etc.

async function generateMcpUI(ctx) {
    const mcpPanel = document.getElementById('mcp-panel');
    if (!mcpPanel) return;

    let container = mcpPanel.querySelector('.mcp-configurations');
    if (!container) {
        mcpPanel.innerHTML = `<h2>MCP Server Configuration</h2>
                              <p>Configure connections to Model-Context-Protocol (MCP) compatible servers.</p>
                              <div class="mcp-configurations"></div>
                              <button id="add-mcp-server-btn" class="btn">Add MCP Server</button>`;
        container = mcpPanel.querySelector('.mcp-configurations');
    }
    container.innerHTML = '<p>Loading MCP configurations...</p>';

    try {
        const backendConfig = await ctx.api.loadBackendConfig();
        const mcpClients = backendConfig.mcp_clients || {};
        container.innerHTML = '';

        if (Object.keys(mcpClients).length === 0) {
            container.innerHTML = '<p>No MCP servers configured yet. Click "Add MCP Server" to begin.</p>';
        } else {
            for (const serverName in mcpClients) {
                addMcpServerForm(ctx, serverName, container, mcpClients[serverName]);
            }
        }
    } catch (error) {
        console.error('Error loading MCP configurations:', error);
        container.innerHTML = `<p class="error">Error loading MCP configurations: ${error.message}</p>`;
    }
}

function addMcpServerForm(ctx, serverName, container, serverConfig = {}) {
    const serverId = serverName || `new-mcp-${Date.now()}`;
    const formHtml = `
        <div class="mcp-server-form" data-server-id="${serverId}">
            <h4>${serverName ? `Edit Server: ${serverName}` : 'New MCP Server'}</h4>
            <div class="form-group">
                <label for="mcp-${serverId}-name">Server Name:</label>
                <input type="text" id="mcp-${serverId}-name" class="mcp-name" value="${serverName || ''}" placeholder="e.g., my-server">
            </div>
            <div class="form-group">
                <label for="mcp-${serverId}-command">Command:</label>
                <textarea id="mcp-${serverId}-command" class="mcp-command" placeholder="e.g., npx -y @modelcontextprotocol/server-everything">${serverConfig.command || ''}</textarea>
            </div>
            <h5>Environment Variables</h5>
            <div class="mcp-env-vars">
                ${Object.entries(serverConfig.env || {}).map(([key, value]) => `
                    <div class="mcp-env-var-item">
                        <input type="text" class="mcp-env-key" placeholder="Key" value="${key}">
                        <span>=</span>
                        <input type="text" class="mcp-env-value" placeholder="Value" value="${value}">
                        <button type="button" class="btn btn-xs btn-danger remove-mcp-env-btn">&times;</button>
                    </div>
                `).join('')}
            </div>
            <button type="button" class="btn btn-sm add-mcp-env-btn">Add Environment Variable</button>
            <button type="button" class="btn btn-sm btn-danger remove-mcp-server-btn">Remove Server</button>
        </div>
    `;
    container.insertAdjacentHTML('beforeend', formHtml);

    const form = container.lastElementChild;

    // Add env var listener
    form.querySelector('.add-mcp-env-btn').addEventListener('click', () => {
        addMcpEnvVarForm(ctx, form.querySelector('.mcp-env-vars'));
        ctx.modifiedFields.mcp.add(serverId);
    });

    // Remove env var listener
    form.querySelectorAll('.remove-mcp-env-btn').forEach(btn => {
        btn.addEventListener('click', () => {
            btn.closest('.mcp-env-var-item').remove();
            ctx.modifiedFields.mcp.add(serverId);
        });
    });

    // Remove server listener
    form.querySelector('.remove-mcp-server-btn').addEventListener('click', () => {
        if (confirm('Remove this MCP server?')) {
            form.remove();
            ctx.modifiedFields.mcp.add(serverId);
        }
    });

    // Input listeners
    form.querySelectorAll('input, textarea').forEach(input => {
        input.addEventListener('input', (event) => ctx.handleInputChange(event));
    });
}

function addMcpEnvVarForm(ctx, container) {
    const envVarHtml = `
        <div class="mcp-env-var-item">
            <input type="text" class="mcp-env-key" placeholder="Key">
            <span>=</span>
            <input type="text" class="mcp-env-value" placeholder="Value">
            <button type="button" class="btn btn-xs btn-danger remove-mcp-env-btn">&times;</button>
        </div>
    `;
    container.insertAdjacentHTML('beforeend', envVarHtml);

    const envVarItem = container.lastElementChild;

    envVarItem.querySelector('.remove-mcp-env-btn').addEventListener('click', () => {
        envVarItem.remove();
        const serverId = envVarItem.closest('.mcp-server-form').dataset.serverId;
        ctx.modifiedFields.mcp.add(serverId);
    });

    envVarItem.querySelectorAll('input').forEach(input => {
        input.addEventListener('input', (event) => ctx.handleInputChange(event));
    });
}

async function saveMcpConfig(ctx) {
    // If no MCP fields were modified, skip saving
    if (ctx.modifiedFields.mcp.size === 0) {
        return;
    }

    try {
        const backendConfig = await ctx.api.loadBackendConfig();

        if (!backendConfig.mcp_clients) {
            backendConfig.mcp_clients = {};
        }

        const newMcpClients = {};

        document.querySelectorAll('.mcp-server-form').forEach(form => {
            const serverId = form.dataset.serverId;
            const nameInput = form.querySelector('.mcp-name');
            const commandInput = form.querySelector('.mcp-command');

            const name = nameInput ? nameInput.value.trim() : '';
            const command = commandInput ? commandInput.value.trim() : '';

            if (!name || !command) return;

            const env = {};
            form.querySelectorAll('.mcp-env-var-item').forEach(item => {
                const key = item.querySelector('.mcp-env-key').value.trim();
                const value = item.querySelector('.mcp-env-value').value.trim();
                if (key) {
                    env[key] = value;
                }
            });

            newMcpClients[name] = {
                command: command,
                env: env
            };
        });

        backendConfig.mcp_clients = newMcpClients;

        await ctx.api.saveBackendConfig(backendConfig, ctx.config.get('pybridge.api_url'));

        ctx.modifiedFields.mcp.clear();
    } catch (error) {
        console.error('Error updating MCP config:', error);
    }
}

function validateMcpStep(ctx) {
    // MCP servers are optional, so the step is always valid
    return true;
}
